src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,132 @@
+using System;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using System.Numerics;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Globalization;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal sealed class W3CLogger : IDisposable
+    {
+        private readonly W3CLoggerProcessor _messageQueue;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+        private readonly W3CLoggingFields _loggingFields;
+
+        internal W3CLogger(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            _options = options;
+
+            _messageQueue = new W3CLoggerProcessor(_options);
+            _loggingFields = _options.CurrentValue.LoggingFields;
+            LogFileFullName = _messageQueue.FullName;
+        }
+
+        public string? LogFileFullName { get; }
+
+        public void Dispose()
+        {
+            if (!(_messageQueue is null))
+            {
+                _messageQueue.Dispose();
+            }
+        }
+
+        public void Log(IReadOnlyCollection<KeyValuePair<string, object?>> state)
+        {
+            _messageQueue.EnqueueMessage(Format(state));
+        }
+
+        private string Format(IEnumerable<KeyValuePair<string, object?>> state)
+        {
+            // Subtract 1 to account for the "All" flag
+            string[] elements = new string[Enum.GetValues(typeof(W3CLoggingFields)).Length - 1];
```

Comment:
That's tricky since the elements need to be in the same order as the field headers, which the array enforces. StringBuilder indexes by char, not as a list of elements, so it's difficult to insert at the right place